docs/configuration.md
# Configuration

English | [Indonesia](configuration-id.md)

```yaml
# configuration example
version: 1
host: 0.0.0.0
port: 8000
headers: 
  X-Custom-Header: "hi"
allowOrigins: "https://gofiber.io, https://gofiber.net"
timezone: Asia/Jakarta
publicDir: public
proxyHeader: CF-Connecting-IP
routes:
  - path: /
    file: index.html
  - path: /about
    file: about.html
  - path: /myjs
    file: myjavascript.js
```

## Version

The version here is not the application version but the configuration version.

## Host

If we set the `host` configuration with a certain IP then the application can only be accessed with that IP. For example, if we set `host` with IP `192.168.1.1`, then when accessing it we have to use the url <http://192.168.1.1:8000>. When we access it via <http://127.0.0.1:8000> it will get an error `ERR_CONNECTION_REFUSED` in your browser. If we don't set it, by default the application can be accessed via various IPs on your computer.

## Port

If we do not set the `port` configuration then by default the application runs on port `8000`. If using Let's Encrypt then port `80` as default.

## TLS Port

If we do not set the `tlsPort` configuration then by default the application runs on port `443`. This configuration is used when Let's Encrypt is set.

## Headers

This configuration to do custom header on response.

## Allow Origins

To enable cors, set the configuration with the origin you want by simply writing `allowOrigins: "https://example.com"`. If you want multiple origins, separate them separately using commas (`,`) like this `allowOrigins: "https://example.com, https://example.net"`. By default cors is empty.

## Timezone

This configuration is to make the time format on the log with the time zone that we specify. For example, we set `timezone` with `Asia/Jakarta` timezone, then the log format will be like this.

```shell
[2023-01-17T17:34:33+07:00] -  - 200 GET / Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36
```

If not set, it will default to `UTC` timezone.

## Public Directory

In the written configuration `publicDir`, this configuration is to specify the location of the `public` directory. For example, when we use React.js, the build directory is named `build`, we can simply set `publicDir` with the name of the build directory. If not set by default, the directory is `public`.

## Proxy Header

This configuration is used to get the user's IP if the host is running behind a Load Balancer. For example, if we use Cloudflare, usually the user's IP is set to a header with the name `CF-Connecting-IP` then in the `proxyHeader` configuration it is set to `proxyHeader: CF-Connecting-IP`. By default the value is empty.

## Routes

By default if we do not set the `routes` configuration, Feserve will use the following configuration.

```yaml
routes:
  - path: *
    file: index.html
```

With the configuration above, every url and parameter will be responded with an `index.html` file, perfect if we use SPA.

You can also manipulate the url as follows.

```yaml
routes:
  - path: /
    file: index.html
  - path: /about
    file: about.html
  - path: /myjs
    file: myjavascript.js
```

With the configuration above when we access `/about` the file that is responded is `about.html` which in fact we can also access it with `/about.html`. Here we also see from the configuration above that if we access `/myjs` we will get a javascript response file of `myjavascript.js`.

If you want the feserve to be a load balancer with a config like the following:

```yaml
routes:
  - path: "*"
    file: index.html
  - path: /payment
    balancer:
      - http://localhost:3000
      - http://localhost:3001
```

## Let's Encrypt

To use Let's Encrypt add a configuration like the following example:

```yaml
letsencrypt:
  email: feserve@example.com
  domains:
    - example.com
    - www.example.com
  certsPath: certs
```


docs/configuration-id.md
# Configuration

[English](configuration.md) | Indonesia

```yaml
# contoh konfigurasi
version: 1
host: 0.0.0.0
port: 8000
headers: 
  X-Custom-Header: "hi"
allowOrigins: "https://gofiber.io, https://gofiber.net"
timezone: Asia/Jakarta
publicDir: public
proxyHeader: CF-Connecting-IP
routes:
  - path: /
    file: index.html
  - path: /about
    file: about.html
  - path: /myjs
    file: myjavascript.js
```

## Version

Versi disini bukan versi aplikasi melainkan versi konfigurasi.

## Host

Jika kita mengatur konfigurasi `host` dengan IP tertentu maka aplikasi hanya dapat diakses dengan IP tersebut. Misalnya kita setting `host` dengan IP `192.168.1.1`, maka saat mengaksesnya kita harus menggunakan url <http://192.168.1.1:8000>. Ketika kita mengaksesnya melalui <http://127.0.0.1:8000> maka akan mendapatkan error `ERR_CONNECTION_REFUSED` pada browser anda. Jika tidak kita atur, secara default aplikasi bisa diakses melalui berbagai IP yang ada di komputer Anda.

## Port

Jika kita tidak mengatur konfigurasi `port` maka secara default aplikasi berjalan pada port `8000`. Jika menggunakan Let's Encrypt maka port `80` sebagai default.

## TLS Port

Jika kita tidak mengatur konfigurasi `tlsPort` maka secara default aplikasi berjalan pada port `443`. Konfigurasi ini digunakan saat Let's Encrypt diatur.

## Headers

Konfigurasi ini untuk melakukan kustom header pada response.

## Allow Origins

Untuk mengaktifkan cors, atur konfigurasi dengan origin yang Anda inginkan cukup dengan menulis `allowOrigins: "https://example.com"`. Jika Anda menginginkan beberapa origin, pisahkan secara terpisah menggunakan koma (`,`) seperti `allowOrigins: "https://example.com, https://example.net"` ini. Secara default cors kosong.

## Timezone

Konfigurasi ini untuk membuat format waktu pada log dengan zona waktu yang kita tentukan. Misal kita setting `timezone` dengan timezone `Asia/Jakarta`, maka format lognya akan seperti ini.

```shell
[2023-01-17T17:34:33+07:00] -  - 200 GET / Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36
```

Jika tidak disetel, ini akan ditetapkan secara default ke zona waktu `UTC`.

## Public Directory

Pada konfigurasi tertulis `publicDir`, konfigurasi ini untuk menentukan lokasi direktori `public`. Misalnya, saat kita menggunakan React.js, direktori build diberi nama `build`, kita cukup menyetel `publicDir` dengan nama direktori build. Jika tidak disetel secara default, direktorinya adalah `public`.

## Proxy Header

Konfigurasi ini digunakan untuk mendapatkan IP pengguna jika host berjalan di belakang Load Balancer. Sebagai contoh, jika kita menggunakan Cloudflare, biasanya IP pengguna diset ke header dengan nama `CF-Connecting-IP` kemudian di konfigurasi `proxyHeader` diset ke `proxyHeader: CF-Connecting-IP`. Secara default nilainya kosong.

## Routes

Secara default jika kita tidak mengatur konfigurasi `routes`, Feserve akan menggunakan konfigurasi berikut.

```yaml
routes:
  - path: *
    file: index.html
```

Dengan konfigurasi di atas, setiap url dan parameter akan direspons dengan file `index.html`, cocok jika kita menggunakan SPA.

Anda juga dapat memanipulasi url sebagai berikut.

```yaml
routes:
  - path: /
    file: index.html
  - path: /about
    file: about.html
  - path: /myjs
    file: myjavascript.js
```

Dengan konfigurasi diatas ketika kita mengakses `/about` file yang direspon adalah `about.html` yang sebenarnya kita juga bisa mengaksesnya dengan `/about.html`. Di sini kita juga melihat dari konfigurasi di atas bahwa jika kita mengakses `/myjs` kita akan mendapatkan file respon javascript dari `myjavascript.js`.

Jika ingin feserve menjadi load balancer dengan config seperti berikut:

```yaml
routes:
  - path: "*"
    file: index.html
  - path: /payment
    balancer:
      - http://localhost:3000
      - http://localhost:3001
```

## Let's Encrypt

Untuk menggunakan Let's Encrypt tambahkan konfigurasi seperti contoh berikut:

```yaml
letsencrypt:
  email: feserve@example.com
  domains:
    - example.com
    - www.example.com
  certsPath: certs
```


internal/config/config.go
package config

import (
	"fmt"
	"os"
	"sync"

	"github.com/spf13/viper"
)

type Config struct {
	Version      string            `yaml:"version"`
	Host         string            `yaml:"host"`
	Port         string            `yaml:"port"`
	TLSPort      string            `yaml:"tlsPort"`
	Headers      map[string]string `yaml:"headers"`
	AllowOrigins string            `yaml:"allowOrigins"`
	TimeZone     string            `yaml:"timezone"`
	PublicDir    string            `yaml:"publicDir"`
	ProxyHeader  string            `yaml:"proxyHeader"`
	Routes       []Routes          `yaml:"routes"`
	Letsencrypt  *Letsencrypt      `yaml:"letsencrypt,omitempty"`
}

type Routes struct {
	Path     string   `yaml:"path"`
	File     string   `yaml:"file"`
	Rewrite  bool     `yaml:"rewrite"`
	Domain   string   `yaml:"domain"`
	Balancer []string `yaml:"balancer"`
}

type Letsencrypt struct {
	Email     string   `yaml:"email"`
	Domains   []string `yaml:"domains"`
	CertsPath string   `yaml:"certsPath"`
}

const version = "1"

var (
	config     *Config
	configOnce sync.Once
)

func Load() *Config {
	configOnce.Do(func() {
		viper.AddConfigPath(".")
		viper.SetConfigName("app")
		_ = viper.ReadInConfig()
		viper.AutomaticEnv()

		config = &Config{}
		if err := viper.Unmarshal(config); err != nil {
			panic(err)
		}

		if _, err := os.Stat("app.yaml"); err == nil {
			if config.Version != version {
				panic(fmt.Sprintf("configuration version must be %s\n", version))
			}
		} else {
			config.Version = version
		}

		if viper.GetString("HOST") != "" {
			config.Host = viper.GetString("HOST")
		}

		if viper.GetString("PORT") != "" {
			config.Port = viper.GetString("PORT")
		}

		if viper.GetString("TLS_PORT") != "" {
			config.TLSPort = viper.GetString("TLS_PORT")
		}

		if config.Port == "" && config.Letsencrypt != nil {
			config.Port = "80"
		}

		if config.Port == "" {
			config.Port = "8000"
		}

		if config.TLSPort == "" {
			config.TLSPort = "443"
		}

		if viper.GetString("TZ") != "" {
			config.TimeZone = viper.GetString("TZ")
		}

		if config.TimeZone == "" {
			config.TimeZone = "UTC"
		}

		if config.PublicDir == "" {
			config.PublicDir = "public"
		}

		if len(config.Routes) == 0 {
			config.Routes = append(config.Routes, Routes{
				Path: "*",
				File: "index.html",
			})
		}

		if config.Letsencrypt != nil {
			if config.Letsencrypt.CertsPath == "" {
				config.Letsencrypt.CertsPath = "certs"
			}
		}
	})
	return config
}
